Implementing high resolution timers on ESP32 and ESP8266

nickovs opened the pull request:
This patch allows the `machine.Timer` class to support timing with microsecond resolution (or better, on the ESP32) rather than the current millisecond resolution.

For one-shot timers this simply provides much better resolution when specifying times. For periodic timers it also provides support for representing frequencies that could not be represented before. For example the old implementation would allow for a 111.11Hz cycle (with a 9ms period) or 125Hz (with an 8ms period) but could not offer a 120Hz cycle; this can now be specified with an 8.3333ms period. It also allows for frequencies that exceed 1KHz.

In addition to supporting floating point values for the existing `period` parameter to the `init()` method this patch adds a floating point `freq` parameter to specify the frequency of a timer in Hz.

The patch is backward API compatible.

The new functionality is conditionally enabled using the new `MICROPY_PY_MACHINE_HI_RES_TIMER` port configuration option. If this is not explicitly set in the then `py/mpconfig.h` sets this to the same value as `MICROPY_PY_BUILTINS_FLOAT` and if floating point support is disabled then the feature is automatically disabled. This PR does not currently explicitly set the `MICROPY_PY_MACHINE_HI_RES_TIMER` in either of the supported ports.

nickovs committed:
Implementing high resolution timers on ESP32 and ESP8266

nickovs commented:
It should be noted that in the current ESP32 code the delivery of interrupt events can be delayed when the main thread is sleeping for long periods of time. See #3894 for more details. Until this issue is fixed foreground code on the ESP32 needs to idle using `utime.sleep_us()` rather than `utime.sleep_ms()` to ensure prompt interrupt delivery.

nickovs committed:
Merge remote-tracking branch 'upstream/master' into esp-accurate-timers

nickovs committed:
Reverted lwip submodule to old version.

nickovs commented:
@dpgeorge I'm not sure why Coveralls is saying coverage decreased, since the only file that I changed that Coveralls says it tested (`py/mpconfig.h`) shows as 100% coverage and the report shows 0 files with changed coverage. I think that this might be an artefact of my having updated the `lwip` submodule and then reverted to the version you had been using, and the coverage dropped back when I reverted to the correct version.

dpgeorge commented on ports/esp32/machine_timer.c:
```diff
@@ -130,20 +136,44 @@ STATIC void machine_timer_enable(machine_timer_obj_t *self) {
+#if MICROPY_PY_MACHINE_HI_RES_TIMER
+    if (args[3].u_obj != mp_const_none) {
+        mp_float_t freq;
+        if (mp_obj_get_float_maybe(args[3].u_obj, &freq)) {
```
To minimise code, I'd usually just use `mp_obj_get_float()`.  That will raise an exception if the arg is not a float or int.  Yes it'll be a generic exception ("can't convert <type> to float") but I think that's fair enough, no real need to have a specific exception here.  The only time you'd trigger the exception is if you passed in a list or something silly.

nickovs replied:
Done. For some reason I thought `mp_obj_get_float()` would fail if the input was an `int` but looking at the code this is obviously not the case.

dpgeorge commented on ports/esp32/machine_timer.c:
```diff
@@ -130,20 +136,44 @@ STATIC void machine_timer_enable(machine_timer_obj_t *self) {
+        }
+    } else {
+        mp_float_t period;
+        if (mp_obj_get_float_maybe(args[2].u_obj, &period)) {
```
This could alse be `mp_obj_get_float()`

nickovs replied:
Also done. I also made the same change in the same two places in the esp8266 port.

dpgeorge commented:
Thanks @nickovs for submitting this, it's a good way to make some progress on making a more portable Timer class.

> In addition to supporting floating point values for the existing period parameter to the init() method this patch adds a floating point freq parameter to specify the frequency of a timer in Hz.

The addition of a `freq` parameter is definitely useful and welcome.  But IMO instead of extending `period` to allow a float it would be better to just go ahead and add `period_us`, which could be just an int or allowed to be a float.  `period_us` gives info about the units and makes it clearer in code what is going on, rather than keeping with `period` which is going to continue to cause confusion as to the units being used.

> The new functionality is conditionally enabled using the new MICROPY_PY_MACHINE_HI_RES_TIMER port configuration option.

As a first step I think we can just enable this high-res timer without having a config option.  It would be backwards compatible.  Then as a second step (if needed) allow the same API (eg `freq` param) to build without float support by using `mp_obj_get_int()` instead of `mp_obj_get_float()`.

> I'm not sure why Coveralls is saying coverage decreased...

No problem, sometimes it gets confused.

nickovs committed:
Switched timer argument parsing to use mp_obj_get_float()

nickovs commented:
> The addition of a freq parameter is definitely useful and welcome. But IMO instead of extending period to allow a float it would be better to just go ahead and add period_us, which could be just an int or allowed to be a float. period_us gives info about the units and makes it clearer in code what is going on, rather than keeping with period which is going to continue to cause confusion as to the units being used.

Sure. The fact that the units for the standard argument are ill-specified has always slightly annoyed me. The other option that I though of was to add a `period_ms` parameter and make it just an alias for the regular `period` parameter, thereby saving on code while both being backward-compatible and making the units explicit.

I have one upy style question: what is your preferred way of dealing with over-speficied parameters? The user should only specify one of `freq`, `period` and `period_us`. I've found some cases where an exception is raised, some where there is some implicit precedence to the arguments and a few where there is an explicit precedence. The former takes more code but is more helpful.

> As a first step I think we can just enable this high-res timer without having a config option. It would be backwards compatible. Then as a second step (if needed) allow the same API (eg freq param) to build without float support by using mp_obj_get_int() instead of mp_obj_get_float().

OK. I will remove the config option and just make the code work the best it can when `MICROPY_PY_BUILTINS_FLOAT` is disabled.

dpgeorge commented:
> The other option that I though of was to add a period_ms parameter and make it just an alias for the regular period parameter, thereby saving on code while both being backward-compatible and making the units explicit.

If we had to choose between just one of `period_ms` and `period_us` I'd go for `us` because it allows more precision (without using floats) but still allows a period up to about 1000 seconds while still staying within a small integer (and hence not require heap allocation).  Maybe we need `period_ns` and/or `period` (seconds) and/or `period_ms` for convenience, but for starters I think `us` is a good option.

> I have one upy style question: what is your preferred way of dealing with over-speficied parameters?

Trying to stay minimal, I'd prefer to just do an implicit precedence without any errors (unless none of them are specified).  Since they are keyword-only args, the user is really going out of their way to provide both and so hopefully sees that it's not right to do this.

dhalbert commented:
Have you all seen the accepted PEP https://www.python.org/dev/peps/pep-0564/, which adds various nanosecond-resolution time functions, including `time.sleep_ns()` ?. It's implemented in Python 3.7.

nickovs committed:
Implemented period_us as requested by @dpgeorge

nickovs commented:
OK, I have added support for a `period_us` argument as discussed, I've removed the `MICROPY_PY_MACHINE_HI_RES_TIMER` config parameter and I've also added support for the `freq` and `period_us` arguments when floating point is not enabled.

Regarding support in the absence of floating point code, while this code is very simple and looks good to me, I have not tested it on the ESP32 port because that port currently fails to build without floating point support (it fails when it tries to compile `modsocket.c`). Note that trying to build the ESP8266 port on `master` without floating point also currently fails because of #3914 but I've submitted a separate PR to fix that (and tested my code using that patch).

dpgeorge commented:
> Have you all seen the accepted PEP https://www.python.org/dev/peps/pep-0564/, which adds various nanosecond-resolution time functions?. It's implemented in Python 3.7.
> 
> (edit: I was wrong about time.sleep_ns() existing, sorry -- wishful thinking). So it only provides a model for dealing with nanosecond values).

Right.  It does provide `time.time_ns()` but this is specified to return an integer and so in uPy will very quickly overflow a small int, requiring allocation on the heap, which we aim to avoid in these core functions.

dpgeorge commented:
> OK. I'm going to add a period_us option and allow it to take a float if supported rather than adding another period_ns option. This will save a little space but provide the flexibility that is needed.

We do try hard to make float really optional, for a few reasons: 1) it costs a lot of code space and some ports cannot afford it (eg cc3200); 2) a minimal port may not want to have floats and we should allow as many of the API functions as possible to still work without floats (in particular for code to be portable to systems without floats); 3) on some ports (eg stm32) floats are allocated on the heap and heap memory is to be avoided where possible (eg for hard interrupts, for determinism, to reduce heap fragmentation).

Eg on stm32, if you want to be able to set a nanosecond (or sub-microsecond) timer in a hard IRQ then you must be able to do it without floats.  So it'll need to have `period_ns`.  Alternatively you could use a large `freq` (but still a small int) to specify sub-microsecond accuracy.  In that case probably an integer `period_us` and `freq` (in Hz) are sufficient to cover most cases and be user friendly.

And it's also worth thinking about spec vs implementation.  If we specify that integer `period_us` and `freq` are the *required* arguments that all timers should implement, but they could optionally take floats, then most scripts should aim to use just integers to be as portable as possible.

From the implementation point of view, on esp8266 it doesn't look like it can do better than microsecond resolution, so there's no real need to allow `period_us` to take floats.  Allowing floats maybe gives the user a false sense of accuracy?

> Do you have any objection to leaving the float support in the existing millisecond period parameter?

Floats are slow on esp8266 because it's all done in software, so from that point of view it is better to avoid them if possible.  So I'd suggest to retain existing integer behaviour for `period`, and also just ints for `period_us` (even `freq` could be constrained to just an int to keep the implementation simple, just using an int to calculate the period for all possible arguments).

dpgeorge commented on ports/esp32/machine_timer.c:
```diff
@@ -36,7 +36,9 @@
 
 #define TIMER_INTR_SEL TIMER_INTR_LEVEL
-#define TIMER_DIVIDER  40000
+#define TIMER_DIVIDER  40
```
If this was "4" could it support 100ns resolution?

nickovs replied:
A value of 4 would give us 50ns resolution (we currently have 500ns resolution). According to the [ESP32 technical reference manual](https://www.espressif.com/sites/default/files/documentation/esp32_technical_reference_manual_en.pdf) (chapter 18) there is a 16-bit prescaler and then a 64-bit counter and the lowest acceptable value for the prescaler is 2. This means that we could have 25ns resolution if we wanted and still have a range of about 14,600 years, which is probably enough. There might be a slight power consumption penalty for doing this since the prescaler does very little but there is a large stack of comparators and whatnot connected to the main counter, but I doubt that it will be measurable in the grand scheme of things.

dpgeorge replied:
Ok, thanks for the details.  Probably 40 (500ns resolution) is a good number because the jitter from other parts of the system (eg caching effects) would drown out higher resolution.  Well, it's easy to tune in the future if needed.

nickovs replied:
I tried switching it to 8 (100ns resolution) and I could'd detect any difference in power consumption. Given how unlikely it is for anyone to leave a micro-controller running for 50,000 years, no matter how good Micropython is, I can't see a reason not to leave it that small.

dpgeorge replied:
I'm happy either way.

dpgeorge commented on ports/esp32/machine_timer.c:
```diff
@@ -130,20 +133,49 @@ STATIC void machine_timer_enable(machine_timer_obj_t *self) {
 
 STATIC mp_obj_t machine_timer_init_helper(machine_timer_obj_t *self, mp_uint_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
     static const mp_arg_t allowed_args[] = {
```
To make the code clearer to read, you may want to add an enum for the args, eg: `enum { ARG_mode, ARG_callback, ARG_period, ARG_freq, ARG_period_us };`, then use them as `args[ARG_freq].u_obj` (this is done elsewhere but it's not consistently used).

nickovs replied:
Good point. I'll do that once we finalise the API.

dpgeorge commented on ports/esp32/machine_timer.c:
```diff
@@ -130,20 +133,49 @@ STATIC void machine_timer_enable(machine_timer_obj_t *self) {
-    self->callback = args[2].u_obj;
+#if MICROPY_PY_BUILTINS_FLOAT
+    if (args[2].u_obj != mp_const_none) {
+	self->period = (uint64_t) (mp_obj_get_float(args[2].u_obj) * TIMER_SCALE / 1000);
```
It may be more accurate if the float is converted to uint64_t first, then the scaling performed....?

nickovs replied:
Not if the period value really is a float. If I'm trying to get 600Hz by specifying a period of 1.6666ms then converting to an uint64_t first will give me 1,000Hz instead.

dpgeorge replied:
Yes, sorry, you're right (I thought this was the "us" calculation).

dpgeorge commented on ports/esp32/machine_timer.c:
```diff
@@ -130,20 +133,49 @@ STATIC void machine_timer_enable(machine_timer_obj_t *self) {
+    } else if (args[4].u_obj != mp_const_none) {
+	self->period = (uint64_t) (mp_obj_get_float(args[4].u_obj) * TIMER_SCALE / 1000000);
+    } else {
+	mp_raise_ValueError("Need period or freq value");
```
Error messages usually start with a lower case letter.

nickovs replied:
Fixed and noted.

nickovs commented:
OK, if you want to try to keep this to only using integers then how would you feel about simply having a parameter for the frequency of the unit in which the period is expressed? For backward compatibility this could default to 1,000 (for milliseconds) but you could set it to 1,000,000 for microseconds or any other value you choose.

This lets us have the period be any rational number that can be expressed with a pair of small integers, which should be sufficiently flexible, and keeps the math simple. For the `freq` parameter I would still be inclined to allow floats if the platform supports them since if you're trying to specify a periodic value then a handful of cycles during the setup doesn't really matter if you're going to leave it running; people who need to set a one shot timer for an accurate time can just provide the rational value with `period` and an appropriate `period_freq` or whatever we call it for the denominator.

dpgeorge commented:
> OK, if you want to try to keep this to only using integers then how would you feel about simply having a parameter for the frequency of the unit in which the period is expressed?

To be clear, the idea is `Timer(period=1, period_hz=1000)` means 1ms, so the actual period in seconds is `period_sec = period / period_hz`?  It's interesting, but, I don't know, it seems a bit obscure and awkward to use/explain.  Would you really want to use these args?  But you are right, it is quite flexible.  `period_hz` is kind of like a prescaler value, and is equivalent to `freq` if `period==1`.

Would `period` and `period_hz` replace `period_us`?

> For the freq parameter I would still be inclined to allow floats if the platform supports

Yes I agree.

nickovs commented:
> To be clear, the idea is `Timer(period=1, period_hz=1000)` means 1ms, so the actual period in seconds is `period_sec = period / period_hz`?

Exactly.

> It's interesting, but, I don't know, it seems a bit obscure and awkward to use/explain. Would you really want to use these args? But you are right, it is quite flexible. period_hz is kind of like a prescaler value, and is equivalent to freq if period==1.

Doing it this way would be a little obscure in the abstract but I think that the concrete examples of `period_hz=1000` and `period_hz=1000000` will make perfect sense to people, as will `period==1, period_hz=something`. There are also cases where you might be trying to implement a protocol where you have some nominal data rate `rate` and you want have the timer go off in `count` bits' time and you can write `period=count, period_hz=rate`; I think this would help make the intent clear and the platform code can work out the best settings for the timer hardware for what you are trying achieve.

> Would period and period_hz replace period_us?

Yes, I think that there would no longer be any need for `period_us` since you just use `period_hz=1000000`.

dpgeorge commented:
> Yes, I think that there would no longer be any need for period_us since you just use period_hz=1000000.

Well, it is a good way to provide a large range of periods specified via integers: with a small int (up to about 10^9) you get a maximum period of 10^9 seconds (32 years), and a minimum period of 1ns, which is a frequency of 1GHz.

Is it then too crazy to repurpose `freq` for the role of `period_hz`?  So the default is 1 for both `Timer(period=1, freq=1)` and the actual period is `period/freq`.  That's not going to be backwards compatible though, and may be a bit confusing that you can specify both `period` and `freq` at the same time (although it does eliminate that problem with precedence of these args, which one gets priority).

nickovs commented:
I don't think that we should use the name `freq` in lieu of `period_hz` since it would be too confusing. The default for `period_hz` needs to be 1000 in order to maintain backward compatibility.

nickovs commented:
In the light of all this discussion I think that the approach in #3922 wins, especially since @dpgeorge has implemented #3923 for the stm32 port. I think that all of the relevant feedback here has been folded into the alternative implementation so I'm closing this PR.
